pkg/querier/frontend2/frontend2.go

```diff
@@ -0,0 +1,290 @@
+package frontend2
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/weaveworks/common/httpgrpc"
+	"github.com/weaveworks/common/user"
+	"go.uber.org/atomic"
+
+	"github.com/cortexproject/cortex/pkg/util/flagext"
+	"github.com/cortexproject/cortex/pkg/util/grpcclient"
+	"github.com/cortexproject/cortex/pkg/util/services"
+)
+
+// Config for a Frontend2.
+type Config struct {
+	SchedulerAddr     string                   `yaml:"scheduler_address"`
+	DNSLookupPeriod   time.Duration            `yaml:"scheduler_dns_lookup_period"`
+	WorkerConcurrency int                      `yaml:"scheduler_worker_concurrency"`
+	GRPCClientConfig  grpcclient.ConfigWithTLS `yaml:"grpc_client_config"`
+
+	// Used to find local IP address, that is sent to scheduler and querier-worker.
+	InfNames []string `yaml:"interface_names"`
+
+	// If set, address is not computed from interfaces.
+	Addr string `yaml:"address" doc:"hidden"`
+	Port int    `doc:"hidden"`
+}
+
+func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
+	f.StringVar(&cfg.SchedulerAddr, "frontend.scheduler-address", "", "DNS hostname used for finding schedulers.")
+	f.DurationVar(&cfg.DNSLookupPeriod, "frontend.scheduler-dns-lookup-period", 10*time.Second, "How often to query DNS.")
+	f.IntVar(&cfg.WorkerConcurrency, "frontend.scheduler-worker-concurrency", 5, "Number of goroutines pushing requests to ")
+
+	cfg.InfNames = []string{"eth0", "en0"}
+	f.Var((*flagext.StringSlice)(&cfg.InfNames), "frontend.interface", "Name of network interface to read address from.")
```

Comment:
I took this from `lifecycler.interface`, where it is called just "interface", but I see we use `instance-interface-names` elsewhere, so that's what I am using here as well now. Thanks.